### PR TITLE
Remove HttpClient as a registered service

### LINKED ISSUE #7585: Implement breaking change: Remove HttpClient as a registered service
Issue tracking change described in: https://github.com/Azure/Azure-Functions/issues/1911

### PR DESCRIPTION
**Pull request by madelinegordon**
Remove HttpClient as a registered service and use HttpClientFactory instead. Added IHttpClientFactory mock to tests for classes whose constructors changed as a result of this change.

### Issue describing the changes in this PR

resolves #7585 

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
   V3 changes (non-breaking) tracked by issue/PR #7674 
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Management/InstanceManager.cs [hunk 1]
```diff
@@ -34,11 +34,11 @@ public class InstanceManager : IInstanceManager
         private readonly HttpClient _client;
         private readonly IScriptWebHostEnvironment _webHostEnvironment;
 
-        public InstanceManager(IOptionsFactory<ScriptApplicationHostOptions> optionsFactory, HttpClient client, IScriptWebHostEnvironment webHostEnvironment,
+        public InstanceManager(IOptionsFactory<ScriptApplicationHostOptions> optionsFactory, IHttpClientFactory clientFactory, IScriptWebHostEnvironment webHostEnvironment,
             IEnvironment environment, ILogger<InstanceManager> logger, IMetricsLogger metricsLogger, IMeshServiceClient meshServiceClient, IRunFromPackageHandler runFromPackageHandler,
             IPackageDownloadHandler packageDownloadHandler)
         {
-            _client = client ?? throw new ArgumentNullException(nameof(client));
+            _client = clientFactory.CreateClient() ?? throw new ArgumentNullException(nameof(clientFactory));
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Management/InstanceManager.cs]: If `clientFactory` is `null` here, you'll get an exception.

Use : `clientFactory?.CreateClient()` instead

File: src/WebJobs.Script.WebHost/Management/LinuxSpecialization/ManagedIdentityTokenProvider.cs [hunk 2]
```diff
@@ -20,11 +20,11 @@ public class ManagedIdentityTokenProvider : IManagedIdentityTokenProvider
         private readonly IMetricsLogger _metricsLogger;
         private readonly ILogger<ManagedIdentityTokenProvider> _logger;
 
-        public ManagedIdentityTokenProvider(IEnvironment environment, HttpClient httpClient,
+        public ManagedIdentityTokenProvider(IEnvironment environment, IHttpClientFactory httpClientFactory,
             IMetricsLogger metricsLogger, ILogger<ManagedIdentityTokenProvider> logger)
         {
             _environment = environment;
-            _httpClient = httpClient;
+            _httpClient = httpClientFactory.CreateClient();
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Management/LinuxSpecialization/ManagedIdentityTokenProvider.cs]: Do we need to create the client this early in all cases? Could the client creation be deferred here?

File: src/WebJobs.Script.WebHost/Management/LinuxSpecialization/PackageDownloadHandler.cs [hunk 3]
```diff
@@ -26,11 +26,11 @@ public class PackageDownloadHandler : IPackageDownloadHandler
         private readonly ILogger<PackageDownloadHandler> _logger;
         private readonly IMetricsLogger _metricsLogger;
 
-        public PackageDownloadHandler(HttpClient httpClient, IManagedIdentityTokenProvider managedIdentityTokenProvider,
+        public PackageDownloadHandler(IHttpClientFactory httpClientFactory, IManagedIdentityTokenProvider managedIdentityTokenProvider,
             IBashCommandHandler bashCommandHandler, ILogger<PackageDownloadHandler> logger,
             IMetricsLogger metricsLogger)
         {
-            _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
+            _httpClient = httpClientFactory.CreateClient() ?? throw new ArgumentNullException(nameof(httpClientFactory));
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Management/LinuxSpecialization/PackageDownloadHandler.cs]: Same comment as above on the `null` case and deferral

File: src/WebJobs.Script.WebHost/Management/MeshServiceClient.cs [hunk 4]
```diff
@@ -27,9 +27,9 @@ public class MeshServiceClient : IMeshServiceClient
 
         private static readonly JsonSerializer Serializer = JsonSerializer.Create(new JsonSerializerSettings());
 
-        public MeshServiceClient(HttpClient client, IEnvironment environment, ILogger<MeshServiceClient> logger)
+        public MeshServiceClient(IHttpClientFactory httpClientFactory, IEnvironment environment, ILogger<MeshServiceClient> logger)
         {
-            _client = client ?? throw new ArgumentNullException(nameof(client));
+            _client = httpClientFactory.CreateClient() ?? throw new ArgumentNullException(nameof(httpClientFactory));
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Management/MeshServiceClient.cs]: same comment as above.

File: src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs [hunk 5]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the MIT License. See License.txt in the project root for license information.
 
 using System;
+using System.Net.Http;
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs]: Is this required?

File: src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs [hunk 6]
```diff
@@ -28,11 +28,11 @@ public class WebFunctionsManager : IWebFunctionsManager
         private readonly HostNameProvider _hostNameProvider;
         private readonly IFunctionMetadataManager _functionMetadataManager;
 
-        public WebFunctionsManager(IOptionsMonitor<ScriptApplicationHostOptions> applicationHostOptions, ILoggerFactory loggerFactory, HttpClient client, ISecretManagerProvider secretManagerProvider, IFunctionsSyncManager functionsSyncManager, HostNameProvider hostNameProvider, IFunctionMetadataManager functionMetadataManager)
+        public WebFunctionsManager(IOptionsMonitor<ScriptApplicationHostOptions> applicationHostOptions, ILoggerFactory loggerFactory, IHttpClientFactory clientFactory, ISecretManagerProvider secretManagerProvider, IFunctionsSyncManager functionsSyncManager, HostNameProvider hostNameProvider, IFunctionMetadataManager functionMetadataManager)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs:31]: nit: recommend you be consistent with your naming. Half of the params are named clientFactory, and half are named httpClientFactory

**Review by brettsam — APPROVED** [review]

File: src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs (hunk 6 above)
Comment (madelinegordon) [reply src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs:31]: Saw this comment just after merging - following up with a separate PR. Thanks @mathewc !

**Pull request merged**